File: {{cookiecutter.project_slug}}/src/auth/apiKey.ts
{% if cookiecutter.sdk_choice == 'typescript' and cookiecutter.auth_mechanism == 'api_key' -%}
/**
 * API Key authentication handler
 * Note: OAuth 2.1 is recommended for public clients
 */

import { Request } from "express";
import * as dotenv from "dotenv";

dotenv.config();

const log = {
  info: (msg: string) => console.error(`[INFO] ${msg}`),
  warn: (msg: string) => console.error(`[WARN] ${msg}`),
};

export class APIKeyHandler {
  private validKeys: Set<string>;

  constructor(validKeys?: string[]) {
    if (validKeys) {
      this.validKeys = new Set(validKeys);
    } else {
      // Load from environment variable
      const keysStr = process.env.MCP_API_KEYS || "";
      this.validKeys = new Set(
        keysStr.split(",").map((k) => k.trim()).filter((k) => k)
      );
    }

    if (this.validKeys.size === 0) {
      log.warn("No API keys configured - all requests will be rejected");
    }
  }

  async authenticate(req: Request): Promise<boolean> {
    // Try Authorization header
    const authHeader = req.headers.authorization;
    if (authHeader && authHeader.startsWith("Bearer ")) {
      const apiKey = authHeader.substring(7);
      if (this.validKeys.has(apiKey)) {
        log.info("Request authenticated via Authorization header");
        return true;
      }
    }

    // Try query parameter
    const queryKey = req.query.api_key as string;
    if (queryKey && this.validKeys.has(queryKey)) {
      log.info("Request authenticated via query parameter");
      return true;
    }

    // Try custom header
    const headerKey = req.headers["x-api-key"] as string;
    if (headerKey && this.validKeys.has(headerKey)) {
      log.info("Request authenticated via X-API-Key header");
      return true;
    }

    log.warn("Authentication failed: Invalid or missing API key");
    return false;
  }

  addKey(apiKey: string): void {
    this.validKeys.add(apiKey);
    log.info("New API key added");
  }

  revokeKey(apiKey: string): void {
    this.validKeys.delete(apiKey);
    log.info("API key revoked");
  }
}
{%- endif %}

File: {{cookiecutter.project_slug}}/src/{{cookiecutter.project_slug}}/auth/api_key.py
{% if cookiecutter.sdk_choice == 'python' and cookiecutter.auth_mechanism == 'api_key' -%}
"""API Key authentication handler."""

import logging
import os
from typing import Optional, Set
from starlette.requests import Request
from dotenv import load_dotenv

logger = logging.getLogger(__name__)

# Load environment variables
load_dotenv()


class APIKeyHandler:
    """API Key authentication handler.

    Note: API keys are simpler but less secure than OAuth 2.1.
    Use OAuth 2.1 for public clients when possible.
    """

    def __init__(self, valid_keys: Optional[Set[str]] = None):
        """Initialize API key handler.

        Args:
            valid_keys: Set of valid API keys. If None, loads from environment.
        """
        if valid_keys is None:
            # Load API keys from environment variable
            api_keys_str = os.getenv('MCP_API_KEYS', '')
            self.valid_keys = set(key.strip() for key in api_keys_str.split(',') if key.strip())
        else:
            self.valid_keys = valid_keys

        if not self.valid_keys:
            logger.warning("No API keys configured. All requests will be rejected.")

    async def authenticate(self, request: Request) -> bool:
        """Authenticate incoming request using API key.

        Supports two methods:
        1. Authorization header: Authorization: Bearer <api_key>
        2. Query parameter: ?api_key=<api_key>

        Args:
            request: Starlette request object

        Returns:
            True if authenticated, False otherwise
        """
        # Try Authorization header first
        auth_header = request.headers.get('Authorization', '')
        if auth_header.startswith('Bearer '):
            api_key = auth_header[7:]  # Remove 'Bearer ' prefix
            if api_key in self.valid_keys:
                logger.info("Request authenticated via Authorization header")
                return True

        # Try query parameter
        api_key = request.query_params.get('api_key')
        if api_key and api_key in self.valid_keys:
            logger.info("Request authenticated via query parameter")
            return True

        # Try custom header
        api_key = request.headers.get('X-API-Key')
        if api_key and api_key in self.valid_keys:
            logger.info("Request authenticated via X-API-Key header")
            return True

        logger.warning("Authentication failed: Invalid or missing API key")
        return False

    def add_key(self, api_key: str) -> None:
        """Add a new valid API key.

        Args:
            api_key: API key to add
        """
        self.valid_keys.add(api_key)
        logger.info("New API key added")

    def revoke_key(self, api_key: str) -> None:
        """Revoke an API key.

        Args:
            api_key: API key to revoke
        """
        self.valid_keys.discard(api_key)
        logger.info("API key revoked")
{%- endif %}
